sound/soc/sof/intel/hda-loader.c

```diff
@@ -551,22 +589,25 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
 	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
 	msg.primary |= SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID(fw_lib->id);
 
-	ret = cl_trigger(sdev, hext_stream, SNDRV_PCM_TRIGGER_START);
-	if (ret < 0) {
-		dev_err(sdev->dev, "%s: DMA trigger start failed\n", __func__);
-		goto cleanup;
-	}
-
+	loader_data.hext_stream = hext_stream;
+	loader_data.dma_started = false;
+	msg.callback_data = &loader_data;
+	msg.callback_after_send = hda_dsp_ipc4_load_library_dma_trigger;
 	ret = sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
 
-	ret1 = cl_trigger(sdev, hext_stream, SNDRV_PCM_TRIGGER_STOP);
-	if (ret1 < 0) {
-		dev_err(sdev->dev, "%s: DMA trigger stop failed\n", __func__);
-		if (!ret)
-			ret = ret1;
+	if (loader_data.dma_started) {
+		/*
+		 * The DMA was started in hda_dsp_ipc4_load_library_dma_trigger()
+		 * callback during the message sending
```

Comment:
maybe add a phrase here - "have to stop it regardless whether we received a reply from the firmware or not"

Reply:
The DMA has to be stopped if it was started, the comment helps the reader who does not see when it is started in this function and might get confused.